src/varlink.js

```diff
@@ -0,0 +1,97 @@
+
+import cockpit from "cockpit";
+
+const encoder = cockpit.utf8_encoder();
+const decoder = cockpit.utf8_decoder();
+
+/*
+ * Connect to a varlink service.
+ *
+ * Returns a connection object:
+ *
+ * - connection.call(method, parameters) calls "method" (prefixed by the
+ *   interface) with "parameters". Multiple calls can be made without waiting
+ *   for replies, but calls are made sequentially (similar to http). Returns a
+ *   promise that resolves with the return value of the method or a tuple
+ *   (varlinkerror, parameters).
+ *
+ * - connection.close() rejects all pending calls and closes the connection.
+ *
+ * - connection.onclosed is called when the connection is closed.
+ *
+ * https://github.com/varlink/varlink.github.io/wiki
+ */
+function connect(address) {
+    if (!address.startsWith("unix:"))
+        throw new Error("Only unix varlink connections supported");
+
+    let connection = {};
+    let pending = [];
+    let buffer = "";
+
+    let channel = cockpit.channel({
+        unix: address.slice(5),
+        binary: true,
+        payload: "stream",
+        superuser: "require"
+    });
+
+    channel.addEventListener("message", (event, data) => {
+        buffer += decoder.decode(data);
+
+        const chunks = buffer.split("\0");
+        buffer = chunks.pop();
+
+        for (let i = 0; i < chunks.length; i += 1) {
+            const message = JSON.parse(chunks[i]);
+
+            let { resolve, reject } = pending.shift();
+            if (message.error)
+                reject({
+                    error: message.error,
+                    parameters: message.parameters
+                });
+            else
+                resolve(message.parameters);
+        }
+    });
+
+    channel.addEventListener("close", (event, options) => {
+        if (connection.onclosed)
+            connection.onclosed(options.problem);
+    });
+
+    connection.call = function (method, parameters) {
+        parameters = parameters || {};
+
+        const data = encoder.encode(JSON.stringify({ method, parameters }));
+
+        channel.send(data);
+        channel.send([0]);
```

Comment:
Since this now knows about `pending`, this could throw  an exception if you are trying to make a call with the previous call's answer still outstanding? This is not a regression, and a follow-up for this is fine, just in case you feel like it while you are cleaning this up.

Reply:
See above. You can queue as many calls as you'd like, without an error. It's just that they'll be handled in sequence, not in parallel.